flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionDriver.java

```diff
@@ -0,0 +1,314 @@
+	private final NodeCache cache;
+
+	/** ZooKeeper path of the node which stores the current leader information. */
+	private final String leaderPath;
+
+	private final ConnectionStateListener listener = (client, newState) -> handleStateChange(newState);
+
+	private final LeaderElectionEventHandler leaderElectionEventHandler;
+
+	private final FatalErrorHandler fatalErrorHandler;
+
+	private final String leaderContenderDescription;
+
+	@GuardedBy("lock")
+	private volatile boolean running;
+
+	/**
+	 * Creates a ZooKeeperLeaderElectionDriver object.
+	 *
+	 * @param client Client which is connected to the ZooKeeper quorum
+	 * @param latchPath ZooKeeper node path for the leader election latch
+	 * @param leaderPath ZooKeeper node path for the node which stores the current leader information
+	 * @param leaderElectionEventHandler Event handler for processing leader change events
+	 * @param fatalErrorHandler Fatal error handler
+	 * @param leaderContenderDescription Leader contender description
+	 */
+	public ZooKeeperLeaderElectionDriver(
+			CuratorFramework client,
+			String latchPath,
+			String leaderPath,
+			LeaderElectionEventHandler leaderElectionEventHandler,
+			FatalErrorHandler fatalErrorHandler,
+			String leaderContenderDescription) throws Exception {
+		this.client = checkNotNull(client);
+		this.leaderPath = checkNotNull(leaderPath);
+		this.leaderElectionEventHandler = checkNotNull(leaderElectionEventHandler);
+		this.fatalErrorHandler = checkNotNull(fatalErrorHandler);
+		this.leaderContenderDescription = checkNotNull(leaderContenderDescription);
+
+		leaderLatch = new LeaderLatch(client, checkNotNull(latchPath));
+		cache = new NodeCache(client, leaderPath);
+
+		client.getUnhandledErrorListenable().addListener(this);
+
+		leaderLatch.addListener(this);
+		leaderLatch.start();
+
+		cache.getListenable().addListener(this);
+		cache.start();
+
+		client.getConnectionStateListenable().addListener(listener);
+
+		running = true;
+	}
+
+	@Override
+	public void close() throws Exception{
+		synchronized (lock) {
+			if (!running) {
+				return;
+			}
+			running = false;
+
+			LOG.info("Closing {}", this);
+
+			client.getUnhandledErrorListenable().removeListener(this);
+
+			client.getConnectionStateListenable().removeListener(listener);
+
+			Exception exception = null;
+
+			try {
+				cache.close();
+			} catch (Exception e) {
+				exception = e;
+			}
+
+			try {
+				leaderLatch.close();
+			} catch (Exception e) {
+				exception = ExceptionUtils.firstOrSuppressed(e, exception);
+			}
+
+			if (exception != null) {
+				throw new Exception("Could not properly stop the ZooKeeperLeaderElectionDriver.", exception);
+			}
+		}
+	}
+
+	@Override
+	public boolean hasLeadership() {
+		return leaderLatch.hasLeadership();
+	}
+
+	@Override
+	public void isLeader() {
+		synchronized (lock) {
+			if (running) {
+				leaderElectionEventHandler.onGrantLeadership();
+			}
+		}
```

Comment:
I have noticed the deadlock problem and moved `leaderElectionDriver.close()` out of the outer lock. This will solve the deadlock problem now. But it is fragile and not very easy to understand. Maybe in the future we introduce some other interfaces, which will be called in the outer lock, then we still have the potential deadlock.

Not execute the callback under the inner lock could completely avoid the deadlock. Just like you say, the implementation(e.g. `DefaultLeaderElection`) needs to filter out the spurious callbacks. This is easier to understand.

Reply:
We could use the `lock` in DefaultLeaderElectionService to easily filter the spurious callbacks. When the `leaderElectionDriver` is closed, the `running` is also false. Then we check the `running` in the callback implementation methods(e.g. `onGrantLeadership`, `onRevokeLeadership`, `onLeaderInformationChange`).